### ISSUE TITLE
[remove_item] limitations

### ISSUE THREAD

**Issue by doofus-01** [wesnoth/wesnoth#4689]
### Game and System Information
 - What version of the game are you running?
1.15.2+dev
 - Where did you download it from? (Steam, Mac App Store, our website, built from source, etc.)
built from source
 - If you're using a custom build, what compiler did you use, and what commit did you build at?
scons, d956d262fe9
 - What OS are you running?
Linux

### Describe the bug
[remove_item] works if you specify the image key or location.  It does not filter if you specify the name key or halo key (without an image key or location), effect is to remove everything.  I didn't check if name key allows filtering within a set of successfully filtered items, as #1472 seems to request.

### To Reproduce

1. First..
Put something like this in a scenario prestart, like 01_The_Elves_Besieged.cfg of HttT
```
        [item]
            halo=items/sword.png
            x,y=18,22
            name=swordy
        [/item]
        [item]
            image=items/sword.png
            x,y=19,22
            name=swordy
        [/item]
        [item]
            image=items/staff.png
            x,y=21,22
            name=staffy
        [/item]
        [item]
            image=items/sword.png~O(0.5)
            x,y=22,20
            name=swordy_a
        [/item]
        [remove_item]
            halo=items/sword.png
#            image=items/sword.png
#            x,y=19,22
#            name=swordy
        [/remove_item]
```

2. Then...
Comment/uncomment the various keys in [remove_item], see what images survive
3. Lastly...

### Expected behavior
1.  filtering with the halo key should behave same as image key
2.  filtering with name key alone should work (like the image key)

### Screenshots
N/A
### Savefiles
If applicable, attach savefiles to help explain your problem. (To find savefiles, start Wesnoth, click on the `i` button at the bottom left corner of the title screen, and open the `Saved games` directory.)
N/A
### Additional context
Add any other context (compiler logs, console output, etc) about the problem here.
N/A

**Comment by gfgtdf** [comment]
hmm yes this is not very well documented, basicially the [remove_item]image= key is always used for all of halo,image and name in [item]. I don't know why the wiki says otherwise. So yes `image` and satandardlocation filter tags are the onyl atrubutes supported in [remove_item] to for exampel remove an item defined with name=swordy, you use [remove_item]image=swordy

**Comment by doofus-01** [comment]
The documentation wasn't great, but the problem is with the current behavior.  
```
        [item]
            halo=items/sword.png
            x,y=18,22
            name=swordy
        [/item]
        [item]
            image=items/sword.png
            x,y=19,22
            name=swordy
        [/item]
        [item]
            image=items/staff.png
            x,y=21,22
            name=items/sword.png
#            name=staffy
        [/item]
        [item]
            image=items/sword.png~O(0.5)
            x,y=22,20
            name=swordy_a
        [/item]
        [remove_item]
#            halo=items/sword.png
            image=items/sword.png
#            x,y=19,22
#            name=swordy
        [/remove_item]
```
The only one that doesn't get removed is the thing at 22,20.  How does that make any sense?

**Comment by gfgtdf** [comment]
> The only one that doesn't get removed is the thing at 22,20. How does that make any sense?

well obviously because all the others have either an image,halo or name that is "items/sword.png". You example also doesn't really make sense as i think noone would usually use 'items/sword.png' as a 'name'.

Furthermore I'm not saying that the current behavior is better than the behavior you expected. Just that its not a bug. If you want to change it and have a good suggestion for a way that is backwards compatible you can make a pr for 1.15 .

**Comment by doofus-01** [comment]
> You example also doesn't really make sense as i think noone would usually use 'items/sword.png' as a 'name'.

Right, which means this is indeed a bug, as the current behavior makes no sense.  I'm not demanding that you fix it right now, and if I someday come up with a fix, I'll be sure to make a PR.  Until then, there exists this bug.

**Comment by CelticMinstrel** [comment]
I guess the problem is that it as an image path function, so a direct compare says it's a different image. We might want a substring compare (at least if the name is a path), perhaps as an alternate attribute in `[remove_item]` (`image_match` maybe)?

**Comment by doofus-01** [comment]
Actually, I thought that was the one part of this that actually was intuitive, since the IPF makes it essentially a different image.  Would wildcard matching be a possibility, rather than a new key?

**Comment by soliton-** [comment]
From what I can tell this wiki edit is simply wrong https://wiki.wesnoth.org/index.php?title=InterfaceActionsWML&diff=60161&oldid=60097

I do agree that it's quite confusing that with the addition of the `name` key in `[item]` `[remove_item] image=` is now also checked against it instead of adding a `name` key in `[remove_item]` like the wiki editor probably assumed.

I would check how often `[item] name=` is actually used to decide what to do.

**Comment by soliton-** [comment]
Btw, bonus confusion since the lua code uses the variable name for the value of `[remove_item] image=`,

**Comment by gfgtdf** [comment]
> From what I can tell this wiki edit is simply wrong

Yes

> I do agree that it's quite confusing that with the addition of the name key in [item]

Yes i agree , i'm not sure whether i said it already here, but the main (only) reason why it was add this way is that halo= is also handled that way (removed by image= and not halo=)

**Comment by gfgtdf** [comment]
Updated the wiki page for now

**Comment by soliton-** [comment]
But for halo it makes sense since it’s also an image. With name there is no connection at all.

I suggest adding a `name` and perhaps a `halo` key to `[remove_item]` and then see what can be done about the `image` key handling.

**Comment by doofus-01** [comment]
I just tripped over this again, so bumping the issue.  `[item]halo=` needing to be removed with `[remove_item]image=` will discourage macro use, but it's hard to see any other upside.  It's pretty counterintuitive, I'd say it's a bug.

**Comment by soliton-** [comment]
Unlike my comment 5 years ago I would now suggest a new API. The interface of `[remove_item]` and `[item]` in general leave much to be desired (the names are already misleading). Leaving the existing tags as they are and adding a new API would be better than breaking compatiblity with a bugfix. My assumption being that a bugfix not only adds support for `halo=` but also removes the halo matching from `image=`. (Which would leave `image=` matching still a mess.)

**Comment by mentos987** [comment]
A new one called [remove_items] , with an added "s", would be good. Have it pass a wml table that looks like the one that created the item. That way you can simply copy/paste the code from the creation into the removal. And have it remove all items that match the input.

**Comment by soliton-** [comment]
I would not mention items at all. While adding image overlays may be part of an item system there is really no other connection.

For a new API something like `[image_overlays]` or similar would make more sense. Removing should definitely be symmetric to creation and there should be no unexpected automatic id/name generation.

Not sure if there is a need for differentiating halo and image. If so there are probably better terms for that as well. But if it's just about at what layer the images are drawn then that can just be another attribute.

**Comment by gfgtdf** [comment]
i agree in particular on the part about [item] making it sounds like its connected to an item system, so something like image_overlays is better.

For removing items, i really want to encourage removing by id/name though, removing by image (or even worse, by location only) always has the risk to randomly remove other addons overlays on that tile that happen to use the same image for something.

**Comment by mentos987** [comment]
> Not sure if there is a need for differentiating halo and image. If so there are probably better terms for that as well. But if it's just about at what layer the images are drawn then that can just be another attribute.

From working on this lately I can say that halos and images do not get along well at all. They currently need to exist on two separate layers, with halos on top (rendered last). 

Images are all perfect hex shapes, and their code assumes this for all updating. Halos are not trimmed to hex size and will thus mess up images if you try to put halos on lower drawing layers. 

One could think about rewriting the image system but I think we should leave it be, it is a good system for the terrain level hex grid. And halos is a good system for larger things that can happen above the terrain, multi-hex animations etc.

**Comment by doofus-01** [comment]
My main issue was that the keys were not the same for [item] and [remove_item], and (at least back when I first posted this) the "name" key was not analogous to the "id" key in other tags - it was not enough by itself, it needed either the location or image.  I haven't tested that part recently.

No problem with a new tag that makes more sense than the legacy [item].  Could the existing macros be updated to use that, for backwards compatibility?

**Comment by soliton-** [comment]
Would have to look at the macros but it seems unlikely in general to be possible especially because of backwards compatibility. Better would be to let UMC authors opt-in to new tags/macros explicitly.
